### PR TITLE
Use String Intern Cache when Constructing DiscoveryNode

### PR DESCRIPTION
**Pull request by original-brownbear**
Use string intern cache when constructing discovery node instances.
These started showing up in profiling now here and there, likely
due to the fact that we're interning more strings via setting string
interning now. Using the deduplicator makes them disappear from profiling again
for the most part.
Also, fix the fact that we would read a mutable `attributes` map if there's
any attributes and escape it via a getter and intern the attribute keys and values
from the wire as well because it's pretty much free to do so now anyway.

The better fix here is probably to just use the `DiscoveryNode` instances less often and just work with node ids instead as they are quite expensive to send over the wire.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/cluster/node/DiscoveryNode.java [hunk 1]
```diff
@@ -216,22 +219,22 @@ public DiscoveryNode(
...
-        this.hostAddress = hostAddress.intern();
+        this.hostAddress = nodeStringDeduplicator.deduplicate(hostAddress);
         this.address = address;
         if (version == null) {
             this.version = Version.CURRENT;
         } else {
             this.version = version;
         }
-        this.attributes = Collections.unmodifiableMap(attributes);
+        this.attributes = Map.copyOf(attributes);
```
Comment (henningandersen) [inline server/src/main/java/org/elasticsearch/cluster/node/DiscoveryNode.java:237]: I am not sure I follow why we want to copy the map here and when deserializing? Wrapping seems less intrusive.

**Review by henningandersen — COMMENTED** [review]
This looks fine though I have one question on it.

**Comment by original-brownbear** [comment]
Jenkins run elasticsearch-ci/part-2 (known and unrelated)

File: server/src/main/java/org/elasticsearch/cluster/node/DiscoveryNode.java (hunk 1 above)
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/cluster/node/DiscoveryNode.java:237]: Hmm yea this isn't too helpful. I figured this map is often just a single element so this is about as expensive as the wrap call but it doesn't really matter -> reverted :)

Comment (henningandersen) [reply server/src/main/java/org/elasticsearch/cluster/node/DiscoveryNode.java:237]: I think we frequently see around 5-6 entries there so could matter a tiny bit.

**Review by henningandersen — APPROVED** [review]
LGTM.

**Comment by original-brownbear** [comment]
Thanks Henning!

**Pull request merged**